tests/Unit/Http/Controller/ApplicationControllerTest.php

```diff
@@ -0,0 +1,228 @@
+        $this->assertEquals('candidate', $controller->getName());
+        $this->assertArrayHasKey('job', $controller->getData());
+        $this->assertArrayHasKey('users', $controller->getData());
+    }
+
+    public function test_showListCandidateApply_not_found_job()
+    {
+        $user = factory(User::class)->make();
+        $user->id = 1;
+        $user->role_id = 2;
+        $this->be($user);
+        $this->expectException(ModelNotFoundException::class);
+        $this->applicationMock->shouldReceive('getJob')->with(10)->andThrow(ModelNotFoundException::class);
+        $this->applicationMock->shouldReceive('showListCandidateApply')->with(10);
+        $controller = $this->applicationController->showListCandidateApply(10);
+        $this->assertEquals('candidate', $controller->getName());
+        $this->assertArrayHasKey('job', $controller->getData());
+        $this->assertArrayHasKey('users', $controller->getData());
+    }
+
+
+    public function test_showHistoryCreateJob_authorized()
+    {
+        $user = factory(User::class)->make();
+        $user->id = 1;
+        $user->role_id = 2;
+        $this->be($user);
+        $jobs = factory(Job::class, 10)->make();
+        $this->applicationMock->shouldReceive('showHistoryCreateJob')->with($user)->andReturn($jobs);
+        $controller = $this->applicationController->showHistoryCreateJob();
+        $this->assertEquals('job_history', $controller->getName());
+        $this->assertArrayHasKey('jobs', $controller->getData());
+    }
+
+    public function test_showHistoryCreateJob_unAuthorized()
+    {
+        $user = factory(User::class)->make();
+        $user->id = 1;
+        $user->role_id = 1;
+        $this->be($user);
+        $jobs = factory(Job::class, 10)->make();
+        $this->expectException(AuthorizationException::class);
+        $this->applicationMock->shouldReceive('showHistoryCreateJob')->with($user)->andReturn($jobs);
+        $controller = $this->applicationController->showHistoryCreateJob();
+        $this->assertEquals('job_history', $controller->getName());
+        $this->assertArrayHasKey('jobs', $controller->getData());
+    }
+
+    public function test_acceptOrReject_authorized()
+    {
+        $user = factory(User::class)->make();
+        $company = factory(Company::class)->make();
+        $user->id = 1;
+        $user->role_id = 2;
+        $this->be($user);
+        $job = factory(Job::class)->make();
+        $job->id = 2;
+        $user->setRelation('company', $company);
+        $job->setRelation('company', $company);
+        $this->applicationMock->shouldReceive('find')->with($job->id)->andReturn($job);
+        $this->applicationMock->shouldReceive('acceptOrReject')->with($job, $user->id, 2);
+        $controller = $this->applicationController->acceptOrReject($user->id, $job->id, 2);
+        $this->assertEquals(route('list_candidate', ['id' => $job->id]), $controller->getTargetUrl());
+    }
+
```

Comment:
giải thích function này

Reply:
- Em fake ra 3 đối tượng user, job,company và set cho user là nhà tuyển dụng với role_id = 2 và đã đăng nhập với hàm $this->be($user).  Sau đó tạo mối quan hệ giữa job và company, user và company để đảm bảo user được fake có quyền update job được fake.
- Dùng hai hàm shouldReceive để chỉnh định hai hàm mà repo sẽ chạy và truyền biến đầu vào và kết quả trả về mong đợi để, sau đó dùng hàm assertEquals() để so đường dẫn trả về mong đợi và đường dẫn trả về của controller